crates/sui-rosetta/src/account.rs

```diff
@@ -10,48 +12,125 @@ use futures::StreamExt;
 use sui_sdk::rpc_types::StakeStatus;
 use sui_sdk::{SuiClient, SUI_COIN_TYPE};
 use sui_types::base_types::SuiAddress;
+use tracing::info;
 
 use crate::errors::Error;
 use crate::types::{
     AccountBalanceRequest, AccountBalanceResponse, AccountCoinsRequest, AccountCoinsResponse,
-    Amount, Coin, SubAccount, SubAccountType, SubBalance,
+    Amount, BlockIdentifier, Coin, SubAccount, SubAccountType, SubBalance,
 };
 use crate::{OnlineServerContext, SuiEnv};
+use std::time::Duration;
 
 /// Get an array of all AccountBalances for an AccountIdentifier and the BlockIdentifier
 /// at which the balance lookup was performed.
 /// [Rosetta API Spec](https://www.rosetta-api.org/docs/AccountApi.html#accountbalance)
-pub async fn balance(
+pub async fn balance_new(
     State(ctx): State<OnlineServerContext>,
     Extension(env): Extension<SuiEnv>,
     WithRejection(Json(request), _): WithRejection<Json<AccountBalanceRequest>, Error>,
 ) -> Result<AccountBalanceResponse, Error> {
     env.check_network_identifier(&request.network_identifier)?;
     let address = request.account_identifier.address;
+    let mut retry_attempts = 5;
     if let Some(SubAccount { account_type }) = request.account_identifier.sub_account {
-        let balances = get_sub_account_balances(account_type, &ctx.client, address).await?;
-        Ok(AccountBalanceResponse {
-            block_identifier: ctx.blocks().current_block_identifier().await?,
-            balances,
-        })
+        while retry_attempts > 0 {
+            let balances_first =
+                get_sub_account_balances(account_type.clone(), &ctx.client, address).await?;
+            let checkpoint1 = ctx
+                .client
+                .read_api()
+                .get_latest_checkpoint_sequence_number()
+                .await?;
+            // Get another checkpoint which is greater than current
+            let mut checkpoint2 = ctx
+                .client
+                .read_api()
+                .get_latest_checkpoint_sequence_number()
+                .await?;
+
+            while checkpoint2 <= checkpoint1 {
+                checkpoint2 = ctx
+                    .client
+                    .read_api()
+                    .get_latest_checkpoint_sequence_number()
+                    .await?;
+                thread::sleep(Duration::from_secs(1))
```

Comment:
These are async functions and you should not use `thread::sleep` instead you should rely on [`tokio::time::sleep`](https://docs.rs/tokio/latest/tokio/time/fn.sleep.html)

Reply:
done